File: app/src/main/java/com/meiji/daily/MainActivity.java
package com.meiji.daily;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.SwitchCompat;
import android.support.v7.widget.Toolbar;
import android.util.TypedValue;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;

import com.afollestad.materialdialogs.Theme;
import com.afollestad.materialdialogs.color.CircleView;
import com.afollestad.materialdialogs.color.ColorChooserDialog;
import com.meiji.daily.di.component.DaggerCommonActivityComponent;
import com.meiji.daily.module.base.BaseActivity;
import com.meiji.daily.module.useradd.UserAddView;
import com.meiji.daily.module.zhuanlan.ZhuanlanView;
import com.meiji.daily.util.RxBusHelper;

import javax.inject.Inject;

import io.reactivex.Flowable;
import io.reactivex.functions.Consumer;

public class MainActivity extends BaseActivity
        implements NavigationView.OnNavigationItemSelectedListener, ColorChooserDialog.ColorCallback {

    static final String TAG = "MainActivity";
    @Inject
    RxBusHelper mRxBusHelper;
    private DrawerLayout mDrawerLayout;
    private NavigationView mNavigationView;
    private SwitchCompat mSwitchCompat;
    private long mExitTime;
    private Flowable<Boolean> mRxBus;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        DaggerCommonActivityComponent.builder()
                .appComponent(App.sAppComponent)
                .build().inject(this);
        super.onCreate(savedInstanceState);
        mRxBus = mRxBusHelper.register(Constant.RxBusEvent.REFRESHUI);
        mRxBus.subscribe(new Consumer<Boolean>() {
            @Override
            public void accept(Boolean isNightMode) throws Exception {
                showAnimation();
                refreshUI();
            }
        });
    }

    @Override
    protected void onDestroy() {
        mRxBusHelper.unregister(Constant.RxBusEvent.REFRESHUI, mRxBus);
        super.onDestroy();
    }

    @Override
    protected int attachLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    protected void initData(Bundle savedInstanceState) {
        replaceFragment(Constant.TYPE_PRODUCT);
        mNavigationView.setCheckedItem(R.id.nav_product);
    }

    @Override
    protected void initViews() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        initToolBar(toolbar, false, null);

        mDrawerLayout = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, mDrawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        mDrawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        mNavigationView = findViewById(R.id.nav_view);
        mNavigationView.setNavigationItemSelectedListener(this);

        mSwitchCompat = mNavigationView.getMenu().findItem(R.id.app_bar_switch).getActionView().findViewById(R.id.switch_input);
        mSwitchCompat.setChecked(mSettingHelper.getIsNightMode());

        setUpSwitch();

        mSwitchCompat.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean isNightMode) {
                mSettingHelper.setIsNightMode(isNightMode);
                setUpSwitch();
                if (isNightMode) {
                    setTheme(R.style.DarkTheme);
                } else {
                    setTheme(R.style.LightTheme);
                }
                mRxBusHelper.post(Constant.RxBusEvent.REFRESHUI, isNightMode);
            }
        });
    }

    private void setUpSwitch() {
        boolean isNightMode = mSettingHelper.getIsNightMode();
        if (isNightMode) {
            mSwitchCompat.setThumbTintList(ColorStateList.valueOf(mSettingHelper.getColor()));
        } else {
            Resources.Theme theme = getTheme();
            Resources resources = getResources();
            TypedValue textColorPrimary = new TypedValue();
            theme.resolveAttribute(R.attr.textColorPrimary, textColorPrimary, true);
            mSwitchCompat.setThumbTintList(resources.getColorStateList(textColorPrimary.resourceId));
        }
    }

    @Override
    public void onBackPressed() {
        long currentTime = System.currentTimeMillis();
        if (mDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            mDrawerLayout.closeDrawer(GravityCompat.START);
        } else if ((currentTime - mExitTime) < 2000) {
            super.onBackPressed();
        } else {
            Snackbar.make(mDrawerLayout, getString(R.string.double_click_exit), Snackbar.LENGTH_SHORT).show();
            mExitTime = currentTime;
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        if (id == R.id.nav_product) {
            replaceFragment(Constant.TYPE_PRODUCT);

        } else if (id == R.id.nav_life) {
            replaceFragment(Constant.TYPE_LIFE);

        } else if (id == R.id.nav_music) {
            replaceFragment(Constant.TYPE_MUSIC);

        } else if (id == R.id.nav_emotion) {
            replaceFragment(Constant.TYPE_EMOTION);

        } else if (id == R.id.nav_profession) {
            replaceFragment(Constant.TYPE_FINANCE);

        } else if (id == R.id.nav_zhihu) {
            replaceFragment(Constant.TYPE_ZHIHU);

        } else if (id == R.id.nav_user_add) {
            replaceFragment(Constant.TYPE_USERADD);

        } else if (id == R.id.nav_color_chooser) {
            createColorChooserDialog();

        } else if (id == R.id.nav_about) {
            AboutActivity.start(this);

        }

        mDrawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    protected void refreshUI() {
        Resources.Theme theme = getTheme();
        TypedValue rootViewBackground = new TypedValue();
        TypedValue textColorPrimary = new TypedValue();
        theme.resolveAttribute(R.attr.rootViewBackground, rootViewBackground, true);
        theme.resolveAttribute(R.attr.textColorPrimary, textColorPrimary, true);
        Resources resources = getResources();
        mNavigationView.setBackgroundResource(rootViewBackground.resourceId);
        mNavigationView.setItemBackgroundResource(rootViewBackground.resourceId);
        mNavigationView.setItemTextColor(resources.getColorStateList(textColorPrimary.resourceId));
        mNavigationView.setItemIconTintList(resources.getColorStateList(textColorPrimary.resourceId));
    }

    private void replaceFragment(int type) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

        Fragment fragmentById = fragmentManager.findFragmentByTag(String.valueOf(type));
        if (fragmentById == null) {
            Fragment fragment;
            if (type != Constant.TYPE_USERADD) {
                fragment = ZhuanlanView.newInstance(type);
            } else {
                fragment = new UserAddView();
            }
            fragmentTransaction
                    .add(R.id.content_main, fragment, String.valueOf(type));
        } else {
            for (Fragment fragment : fragmentManager.getFragments()) {
                fragmentTransaction.hide(fragment);
            }
            fragmentTransaction.show(fragmentById);
        }
        fragmentTransaction.commit();
    }

    private void createColorChooserDialog() {
        new ColorChooserDialog.Builder(this, R.string.md_color_chooser_title)
                .doneButton(R.string.md_done)
                .cancelButton(R.string.md_cancel)
                .backButton(R.string.md_back)
                .allowUserColorInput(true)
                .allowUserColorInputAlpha(false)
                .theme(mSettingHelper.getIsNightMode() ? Theme.DARK : Theme.LIGHT)
                .customButton(R.string.md_custom)
                .show();
    }

    @Override
    public void onColorSelection(@NonNull ColorChooserDialog dialog, @ColorInt int selectedColor) {
        if (getSupportActionBar() != null)
            getSupportActionBar().setBackgroundDrawable(new ColorDrawable(selectedColor));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().setStatusBarColor(CircleView.shiftColorDown(selectedColor));
            getWindow().setNavigationBarColor(selectedColor);
        }
        if (!dialog.isAccentMode()) {
            mSettingHelper.setColor(selectedColor);
        }
    }

    @Override
    public void onColorChooserDismissed(@NonNull ColorChooserDialog dialog) {
        setUpSwitch();
    }

    private void showAnimation() {
        final View decorview = getWindow().getDecorView();
        Bitmap cacheBitmap = getCacheBitmapFromView(decorview);
        if (decorview instanceof ViewGroup && cacheBitmap != null) {
            final View view = new View(this);
            view.setBackground(new BitmapDrawable((getResources()), cacheBitmap));
            ViewGroup.LayoutParams layoutParams = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT);
            ((ViewGroup) decorview).addView(view, layoutParams);
            ObjectAnimator objectAnimator = ObjectAnimator.ofFloat(view, "alpha", 1f, 0f);
            objectAnimator.setDuration(300);
            objectAnimator.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    ((ViewGroup) decorview).removeView(view);
                }
            });
            objectAnimator.start();
        }
    }

    private Bitmap getCacheBitmapFromView(View view) {
        final boolean drawingCacheEnable = true;
        view.setDrawingCacheEnabled(drawingCacheEnable);
        view.buildDrawingCache(drawingCacheEnable);
        final Bitmap drawingCache = view.getDrawingCache();
        Bitmap bitmap;
        if (drawingCache != null) {
            bitmap = Bitmap.createBitmap(drawingCache);
            view.setDrawingCacheEnabled(false);
        } else {
            bitmap = null;
        }
        return bitmap;
    }
}

File: app/src/main/java/com/meiji/daily/MainActivity.kt
package com.meiji.daily

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.support.annotation.ColorInt
import android.support.design.widget.NavigationView
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.widget.SwitchCompat
import android.util.TypedValue
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.Theme
import com.afollestad.materialdialogs.color.CircleView
import com.afollestad.materialdialogs.color.ColorChooserDialog
import com.meiji.daily.di.component.DaggerCommonActivityComponent
import com.meiji.daily.module.base.BaseActivity
import com.meiji.daily.module.useradd.UserAddView
import com.meiji.daily.module.zhuanlan.ZhuanlanView
import com.meiji.daily.util.RxBusHelper
import io.reactivex.Flowable
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.item_switch.view.*
import kotlinx.android.synthetic.main.toolbar.*
import javax.inject.Inject

class MainActivity : BaseActivity(), NavigationView.OnNavigationItemSelectedListener, ColorChooserDialog.ColorCallback {
    @Inject
    lateinit var mRxBusHelper: RxBusHelper
    private lateinit var mSwitchCompat: SwitchCompat
    private var mExitTime: Long = 0
    private var mRxBus: Flowable<Any>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        DaggerCommonActivityComponent.builder()
                .appComponent(App.sAppComponent)
                .build().inject(this)
        super.onCreate(savedInstanceState)
        mRxBus = mRxBusHelper.register(Constant.RxBusEvent.REFRESHUI)
        mRxBus?.subscribe {
            showAnimation()
            refreshUI()
        }
    }

    override fun onDestroy() {
        mRxBusHelper.unregister(Constant.RxBusEvent.REFRESHUI, mRxBus!!)
        super.onDestroy()
    }

    override fun attachLayoutId() = R.layout.activity_main

    override fun initData(savedInstanceState: Bundle?) {
        replaceFragment(Constant.TYPE_PRODUCT)
        nav_view.setCheckedItem(R.id.nav_product)
    }

    override fun initViews() {
        initToolBar(toolbar, false, null)

        ActionBarDrawerToggle(this, drawer_layout, toolbar,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close).let {
            drawer_layout.addDrawerListener(it)
            it.syncState()
        }

        nav_view.setNavigationItemSelectedListener(this)

        mSwitchCompat = nav_view.menu.findItem(R.id.app_bar_switch).actionView.switch_input
        mSwitchCompat.isChecked = mSettingHelper.isNightMode

        setUpSwitch()

        mSwitchCompat.setOnCheckedChangeListener { compoundButton, isNightMode ->
            mSettingHelper.isNightMode = isNightMode
            setUpSwitch()
            when {
                isNightMode -> setTheme(R.style.DarkTheme)
                else -> setTheme(R.style.LightTheme)
            }
            mRxBusHelper.post(Constant.RxBusEvent.REFRESHUI, isNightMode)
        }
    }


    @Suppress("deprecation")
    private fun setUpSwitch() {
        val isNightMode = mSettingHelper.isNightMode
        if (isNightMode) {
            mSwitchCompat.thumbTintList = ColorStateList.valueOf(mSettingHelper.color)
        } else {
            val textColorPrimary = TypedValue()
            theme.resolveAttribute(R.attr.textColorPrimary, textColorPrimary, true)
            mSwitchCompat.thumbTintList = resources.getColorStateList(textColorPrimary.resourceId)
        }
    }

    override fun onBackPressed() {
        val currentTime = System.currentTimeMillis()
        when {
            drawer_layout.isDrawerOpen(GravityCompat.START) -> drawer_layout.closeDrawer(GravityCompat.START)
            (currentTime - mExitTime < 2000) -> super.onBackPressed()
            else -> {
                Snackbar.make(drawer_layout, getString(R.string.double_click_exit), Snackbar.LENGTH_SHORT).show()
                mExitTime = currentTime
            }
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId

        when (id) {
            R.id.nav_product -> replaceFragment(Constant.TYPE_PRODUCT)
            R.id.nav_life -> replaceFragment(Constant.TYPE_LIFE)
            R.id.nav_music -> replaceFragment(Constant.TYPE_MUSIC)
            R.id.nav_emotion -> replaceFragment(Constant.TYPE_EMOTION)
            R.id.nav_profession -> replaceFragment(Constant.TYPE_FINANCE)
            R.id.nav_zhihu -> replaceFragment(Constant.TYPE_ZHIHU)
            R.id.nav_user_add -> replaceFragment(Constant.TYPE_USERADD)
            R.id.nav_color_chooser -> createColorChooserDialog()
            R.id.nav_about -> AboutActivity.start(this)
        }

        drawer_layout.closeDrawer(GravityCompat.START)
        return true
    }

    @Suppress("deprecation")
    private fun refreshUI() {
        val rootViewBackground = TypedValue()
        val textColorPrimary = TypedValue()
        theme.resolveAttribute(R.attr.rootViewBackground, rootViewBackground, true)
        theme.resolveAttribute(R.attr.textColorPrimary, textColorPrimary, true)
        nav_view.setBackgroundResource(rootViewBackground.resourceId)
        nav_view.itemTextColor = resources.getColorStateList(textColorPrimary.resourceId)
        nav_view.itemIconTintList = resources.getColorStateList(textColorPrimary.resourceId)
    }

    private fun replaceFragment(type: Int) {
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()

        val fragmentById = fragmentManager.findFragmentByTag(type.toString())
        if (fragmentById == null) {
            val fragment: Fragment = when {
                type != Constant.TYPE_USERADD -> ZhuanlanView.newInstance(type)
                else -> UserAddView()
            }
            fragmentTransaction.add(R.id.content_main, fragment, type.toString())
        } else {
            fragmentManager.fragments.forEach { fragment -> fragmentTransaction.hide(fragment) }
            fragmentTransaction.show(fragmentById)
        }
        fragmentTransaction.commit()
    }

    private fun createColorChooserDialog() {
        ColorChooserDialog.Builder(this, R.string.md_color_chooser_title)
                .doneButton(R.string.md_done)
                .cancelButton(R.string.md_cancel)
                .backButton(R.string.md_back)
                .allowUserColorInput(true)
                .allowUserColorInputAlpha(false)
                .theme(if (mSettingHelper.isNightMode) Theme.DARK else Theme.LIGHT)
                .customButton(R.string.md_custom)
                .show()
    }

    override fun onColorSelection(dialog: ColorChooserDialog, @ColorInt selectedColor: Int) {
        if (supportActionBar != null)
            supportActionBar?.setBackgroundDrawable(ColorDrawable(selectedColor))
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.statusBarColor = CircleView.shiftColorDown(selectedColor)
            window.navigationBarColor = selectedColor
        }
        if (!dialog.isAccentMode) {
            mSettingHelper.color = selectedColor
        }
    }

    override fun onColorChooserDismissed(dialog: ColorChooserDialog) {
        setUpSwitch()
    }

    private fun showAnimation() {
        val decorView = window.decorView
        val cacheBitmap = getCacheBitmapFromView(decorView)
        if (decorView is ViewGroup && cacheBitmap != null) {
            val view = View(this)
            view.background = BitmapDrawable(resources, cacheBitmap)
            val layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT)
            decorView.addView(view, layoutParams)
            val objectAnimator = ObjectAnimator.ofFloat(view, "alpha", 1f, 0f)
            objectAnimator.duration = 300
            objectAnimator.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    super.onAnimationEnd(animation)
                    decorView.removeView(view)
                }
            })
            objectAnimator.start()
        }
    }

    private fun getCacheBitmapFromView(view: View): Bitmap? {
        val drawingCacheEnable = true
        view.isDrawingCacheEnabled = drawingCacheEnable
        view.buildDrawingCache(drawingCacheEnable)
        val drawingCache = view.drawingCache
        val bitmap: Bitmap?
        if (drawingCache != null) {
            bitmap = Bitmap.createBitmap(drawingCache)
            view.isDrawingCacheEnabled = false
        } else {
            bitmap = null
        }
        return bitmap
    }

    companion object {
        const val TAG = "MainActivity"
    }
}
